### PR TITLE
FOGL-3027: initial check-in of simple python filter plugin

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-3027: initial check-in of simple python filter plugin

The plugin is currently using "string" item in configuration.

This means we can use only one statement in the inout field.

Adding \n results in an issue at reconfigure time.

Next action would be the creation of a new configuration item, say "code", "text", that will be fine for init and reconfigure.

This new  configuration item will be similar to "script" item which is not escaped by ConfigCategory class internals

Example usage:

1) set randomwalk filter:
2) add the following code:
`reading[b'randomwalk'] = reading[b'randomwalk'] * 10 + 330`

### CODE REVIEW

File: plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+	 *
+	 * Fetch all Datapoins in 'reading' dict
+	 * dKey and dValue are borrowed references
+	 */
+	PyObject *dKey, *dValue;
+	Py_ssize_t dPos = 0;
+	while (PyDict_Next(result, &dPos, &dKey, &dValue))
+	{
+		DatapointValue* dataPoint;
+		if (PyLong_Check(dValue) || PyLong_Check(dValue))
```
Comment (AmandeepArora) [inline plugin.cpp]: same check twice.

File: plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+		}
+		else
+		{
+			// Get 'reading' value: borrowed reference from localDictionary
+			PyObject* result = PyDict_GetItemString(localDictionary, "reading");
+			/*
+			 * Create a Reading object from Python dict
+			 *
+			 * Fetch all Datapoins in 'reading' dict:
+			 * dKey and dValue are borrowed references
```
Comment (AmandeepArora) [inline plugin.cpp]: dKey and dValue are not being referenced here, they are in helper function getFilteredReadingData, update the comment here.

File: plugin.cpp [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+		"default" :  FILTER_NAME,
+		"readonly": "true"
+		},
+	"enable": {
+		"description": "A switch that can be used to enable or disable execution of the Simple Python filter.",
+		"type": "boolean",
+		"displayName": "Enabled",
+		"default": "false"
+		},
+	"code": {
```
Comment (MarkRiddoch) [inline plugin.cpp]: Please add order property

File: plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+	SimplePythonFilter* handle =
+		new SimplePythonFilter(FILTER_NAME,
+					*config,
+					outHandle,
+					output);
+
+	if (config->itemExists("code"))
+	{
+		handle->m_code = config->getValue("code");
+	}
```
Comment (MarkRiddoch) [inline plugin.cpp]: If code doesn't exist then consider raising an error.

File: plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+
+	if (!Py_IsInitialized())
```
Comment (MarkRiddoch) [inline plugin.cpp]: If this fails then raise an error

File: plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+#ifdef PLUGIN_PYTHON_SHARED_LIBRARY
+		string openLibrary = TO_STRING(PLUGIN_PYTHON_SHARED_LIBRARY);
+		if (!openLibrary.empty())
+		{
+			libpython_handle = dlopen(openLibrary.c_str(),
+						  RTLD_LAZY | RTLD_GLOBAL);
+			Logger::getLogger()->info("Pre-loading of library '%s' "
+						  "is needed on this system",
+						  openLibrary.c_str());
+		}
```
Comment (MarkRiddoch) [inline plugin.cpp]: Failure case? What is the implication? Should it at least be logged?

File: plugin.cpp (hunk 5 above)
Comment (pintomax) [reply plugin.cpp]: We don't need to raise an error: if Python library is already loaded we can continue

**Comment by pintomax** [comment]
Next commits will be about multi-statements support.

After that I'll add all the needed examples

**Comment by pintomax** [comment]
Added support for multi-statements separated by \n

File: plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+	PyErr_Fetch(&pType, &pValue, &pTraceback);
+	PyErr_NormalizeException(&pType, &pValue, &pTraceback);
+
+	PyObject* str_exc_value = PyObject_Repr(pValue);
+	PyObject* pyExcValueStr = PyUnicode_AsEncodedString(str_exc_value,
+							    "utf-8",
+							    "Error ~");
+
+	// NOTE from :
+	// https://docs.python.org/2/c-api/exceptions.html
```
Comment (praveen-garg) [inline plugin.cpp]: https://docs.python.org/3.5/c-api/exceptions.html

File: include/simple_python.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
+#ifndef _SIMPLE_PYTHON35_FILTER_H
```
Comment (praveen-garg) [inline include/simple_python.h]: _SIMPLE_PYTHON35_FILTER_H should be renamed to _SIMPLE_PYTHON3_FILTER_H

As we also support Python3.6 and 3.7 i.e. all 3.5.3+

File: plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+ * The Filter plugin interface
+ */
+extern "C" {
+
+/**
+ * The plugin information structure
+ */
+static PLUGIN_INFORMATION info = {
+	FILTER_NAME,              // Name
+	"1.0.0",                  // Version
```
Comment (praveen-garg) [inline plugin.cpp]: VERSION (from version.h)

File: CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+link_directories(${FOGLAMP_LIB_DIRS})
+
+# Create shared library
+add_library(${PROJECT_NAME} SHARED ${SOURCES} version.h)
+
+# Add FogLAMP library names
+target_link_libraries(${PROJECT_NAME} ${NEEDED_FOGLAMP_LIBS})
+
+# Add additional libraries
+# Add Python 3.5 library
```
Comment (praveen-garg) [inline CMakeLists.txt]: replace all such `3.5` occurrences with 3.x (except the url link in comment for exception handling)

File: include/simple_python.h [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+#define _SIMPLE_PYTHON35_FILTER_H
+/*
+ * FogLAMP "Simple Python 3.5" filter class.
```
Comment (praveen-garg) [inline include/simple_python.h]: 3.x or just python3

File: include/simple_python.h [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+#include <filter_plugin.h>
+#include <filter.h>
+
+#include <Python.h>
+
+/**
+ * Python35Filter class is derived from FogLampFilter
+ * It handles loading of a python module (provided script name)
+ * and the Python3.5 C API calls
```
Comment (praveen-garg) [inline include/simple_python.h]: Python35Filter - SimplePythonFilter
Python3.5 - Python3.x or just Python3

File: README.rst [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+- Create a new datapoint while filtering
+
+.. code-block:: console
+
+    reading[b'temp_fahr'] = reading[b'temperature'] * 9 / 5 + 32
+
+- Generate an exponential moving average (ema)
+   In this case we need to persing some data while filering current dataset
+   the filter receives in input.
+   The global 'user_data' empty dictonary is available to the Python interpreter
```
Comment (praveen-garg) [inline README.rst]: The global > A global

typo: 

dictonary

**Comment by praveen-garg** [comment]
@pintomax  @mshariq-nerd  helped (sanity) testing this. the documentation fixes needs one more iteration, otherwise looks good and working fine !

if you get a chance, to some stress test with RPi0

**Comment by pintomax** [comment]
@praveen-garg tested on Ubuntu 16 /Pi3 and Pi4. Next one will be PiZero

File: README.rst [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+   In this case we need to persing some data while filtering current dataset
+   the filter receives in input.
+   A global 'user_data' empty dictionary is available to the Python interpreter
+   and key values can be easily added:
+
+.. code-block:: console
+
+    global user_data
+    if not user_data:
+    user_data['latest'] = None
```
Comment (praveen-garg) [inline README.rst]: indent w.r.t if

**Comment by pintomax** [comment]
@praveen-garg, just tested three simple filters running in PiZero Buster.
Data generator is "randomwalk" plugin set with 10 readings/s

**Pull request merged**